### DISCUSSION TITLE
Provide info on why certain icons cannot be requested

### DISCUSSION THREAD

**Discussion by moertel** [zixpo/candybar#133] [Ideas]
@sarsamurmu thanks a lot for implementing #123, much appreciated! From what I see, the new filter sits in the place that detects the installed apps: https://github.com/zixpo/candybar/blob/529a7619c4e03e814ddf4430258ff778cf734159/library/src/main/java/candybar/lib/tasks/IconRequestTask.java#L113-L114

Personally, I'd prefer icons to still be listed in the icon request section even if a filter is applied, but maybe greyed out, and display a reason to the user why they can't request it. Otherwise it's easy to look like a bug to the user and cause frustration.

Would you be open to such a contribution?

I'd keep the current interface intact and keep using the `true/false` return value of `filterRequest()` to signify whether the app should be filtered out "hard", i.e. be made invisible. The implementation I have in mind would add two variables to the `Request` class:
 * `boolean isAvailableForRequest` (default `true`, can be written from `FilterRequestHandler`)
 * `String infoText` (default `''`, can be written from `FilterRequestHandler`)

Covering these use-cases:
1. `filterRequest()=false` --> not displayed
2. `filterRequest()=true` and `isAvailableForRequest=false` --> displayed but greyed out
3. `isAvailableForRequest=false` and `infoText='Web apps cannot be requested'` --> greyed out with info
4. `isAvailableForRequest=true` and `infoText='App not listed on GooglePlay'` --> displayed with info

Some screenshots for demo purposes. They use the proposed implementation and the following code:
```java
configuration.setFilterRequestHandler(request -> {
    if (request.getName().contains("Files")) {
        request.setAvailableForRequest(false);
        request.setInfoText("This icon is a web shortcut and not associated with an Android app. Unfortunately it cannot be requested at this time.\n\nIn many launchers, you can long-press the app icon in the drawer and pick an existing icon from the icon pack.");
    }
    return true;
});
```
| Regular view        | View when tapped |
| ----------------- | ------------------ |
| <img src="https://github.com/zixpo/candybar/assets/1823966/83b3c6ee-a720-46b6-892b-bfb9f5029a75" width="300px" /> | <img src="https://github.com/zixpo/candybar/assets/1823966/36ebd7e7-96a0-4bc7-984a-970a7dc45a14" width="300px" /> |

**Comment by moertel** [comment]
@Donnnno FYI, curious what you think about this.

**Reply by Donnnno** [reply]
Looks very nice!

**Comment by moertel** [comment]
Implementation: https://github.com/zixpo/candybar/pull/134

**Reply by Donnnno** [reply]
Great!
This technique can also be used to filter stuff out like lucky patcher right?

**Reply by moertel** [reply]
Absolutely. Although in practice I found it difficult as the name and the package are super variable and often randomised (e.g. `L*cky P*tcher`).

Edit: not to mention _the audacity_ of people to even send such a request. I usually just delete them and move on without guilt.

**Reply by moertel** [reply]
In case you're interested - the way I intend to use it is making light-weight HTTP requests to F-Droid and Google Play and warning people if their app is not listed on either marketplace. I don't block those requests because system apps (gallery, calendar, phone, etc.) are legitimate exceptions but I do want to display a little warning.

```java
configuration.setFilterRequestHandler(request -> {
    String fDroid = "https://f-droid.org/en/packages/" + request.getPackageName();
    String googlePlay = "https://play.google.com/store/apps/details?id=" + request.getPackageName();
    if (!isAvailableOnMarketplace(googlePlay) && !isAvailableOnMarketplace(fDroid)) {
        request.setInfoText("This app is not listed on Google Play.\n\nIf the app is a system app (e.g. camera, phone, etc.) then this is normal and no reason to worry. An icon request may be discarded if:\n\n\t• the app is illegal (e.g. patchers)\n\t• it is NSFW (not safe for work)\n\t• it is nowhere officially released\n\nIf you're unsure before making a request, reach out to me at android@moertel.app and I'll be happy to help.");
    }
    return true;
});
```
This is the method used:
```java
private boolean isAvailableOnMarketplace(String url) {
    boolean isAvailable = false;
    try {
        HttpURLConnection.setFollowRedirects(false);
        HttpURLConnection con = (HttpURLConnection) new URL(url).openConnection();
        con.setRequestMethod("HEAD");
        if (con.getResponseCode() == HttpURLConnection.HTTP_OK) isAvailable = true;
        con.disconnect();
    } catch (Exception ignored) {
        // Do nothing :)
    }
    return isAvailable;
}
```
And looks like this:

<img src="https://github.com/zixpo/candybar/assets/1823966/8d9f4acf-3549-42ce-9413-474f27ce39d9" width="300px"/> <img src="https://github.com/zixpo/candybar/assets/1823966/eb9f82e6-50eb-4f90-aae1-1c9a4e16ebac" width="300px"/>

**Comment by sarsamurmu** [comment]
Yeah, this is good. I didn't really think much while implementing this feature. Thanks for addressing the issue.

**Comment by walrus543** [comment]
I use the current implementation to filter out piracy apps. Many apps can be identified thanks to the "start by" approach.
However many legit apps are neither on Google Play nor F-Droid.

Keeping the related apps in the list with the addition of an explanation is indeed a good idea. Thank you!

Note: I think a quick how-to guide would be appreciated for such new features. I'm afraid several icon creators won't know how to use them (e.g. the recent PR for analytics).

**Reply by moertel** [reply]
> Note: I think a quick how-to guide would be appreciated for such new features. I'm afraid several icon creators won't know how to use them (e.g. the recent PR for analytics).

I'm with you there 100%. Please don't interpret the lack of a how-to guide as a lack of willingness to contribute one. :) I'm a bit in firefighting-mode these days and am fixing the biggest pain points of my audience (and myself). That's why you see code contributions and not much else. Especially the analytics topic is very dear to my heart, though, so I want to add a guide as soon as I get the time.

Meanwhile, I think it's absolutely fine to ask "how do I do X" questions in the `Discussions` section here.

**Comment by Donnnno** [comment]
So i'm implementing this in my packs right now and was wondering where the text stuff goes with the standard webapk stuff:

![image](https://github.com/zixpo/candybar/assets/31142286/b0d97984-00ee-49f8-9150-4184c0df8849)

**Reply by sarsamurmu** [reply]
```java
String pkg = request.getPackageName();
if (pkg == null) return true;
if (pkg.startsWith("<chrome_thing>") || pkg.startsWith("<samsung_thing>")) {
  request.setAvailableForRequest(false);
  request.setInfoText("<the_text>");
}
return true;
```

**Reply by Donnnno** [reply]
Thank you! I just can't understand the logic of java haha

**Comment by walrus543** [comment]
I have a force close when opening the icon request screen (I upgraded to 3.17.4).
Here is my CandyBar.java file:

```
        configuration.setFilterRequestHandler((request) -> {
            // Return true to include the request
            // Return false to exclude the request
            
            String pkg = request.getPackageName();
            if (pkg == null) return true;
            if (pkg.startsWith("com.whatsapp") || pkg.startsWith("com.instagram.android")) {
            request.setAvailableForRequest(false);
            request.setInfoText("<the_text>");
            }
            return true;
        });
```
What's wrong?

**Reply by moertel** [reply]
It would definitely help if you could have a look at your logs while trying this out to see where the force close is coming from. It's hard to guess otherwise. The code looks ok to me, although I can imagine the `<` and `>` in the text potentially causing issues.

**Reply by walrus543** [reply]
All good. The code was fine. I had to update `fragment_request_item_list.xml` since I have a local version.
Thank you.

**Comment by moertel** [comment]
Was implemented with https://github.com/zixpo/candybar/pull/134. Examples can be found in the comments above:
 * Exclude by package name: https://github.com/zixpo/candybar/discussions/133#discussioncomment-6141326
 * Exclude by marketplace: https://github.com/zixpo/candybar/discussions/133#discussioncomment-5957071
